Jenkinsfile: add stage for Windows 2022 (SAC)

thaJeztah opened the pull request:
This adds a stage to test against the current SAC (Semi Annual Channel),
which allows us to catch possible regressions on upcoming LTS versions.

addresses [ENGCORE-832](https://docker.atlassian.net/browse/ENGCORE-832)

thaJeztah committed:
Jenkinsfile: add stage for Windows 1903 (SAC)

This adds a stage to test against the current SAC (Semi Annual Channel),
which allows us to catch possible regressions on upcoming LTS versions.

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
This might show the problem that's reported in https://github.com/docker/for-win/issues/3884, which was indicated to be a regression in the platform; https://github.com/docker/for-win/issues/3884#issuecomment-510973455

> The problem is now understood. It looks like it will require a fix for a regression in docker, and a fix in Windows itself. For now, the only workaround is as others have put here to take out the storage-opt in daemon.json. It also affects docker run where `--storage-opt size=30GB` for example (any number, 30 is just arbitrary for demonstration). It also means that it is not possible to build layers larger than 20GB in Windows 1903 builds. 

But only shows on Docker Desktop (because Docker Enterprise is only supported on Windows Server LTS versions), and thus doesn't show up in current CI in this repository (which didn't test against 1903).

Possibly skipping this step could work around the platform regression; 
https://github.com/moby/moby/blob/facad557440a0c955beb615495b8d0175f25e4e3/builder/dockerfile/internals.go#L458-L465

thaJeztah commented:
The good news; Windows 1903 machines are working, and CI starts;

```
Client:
 Debug Mode: false
 Plugins:
  cluster: Manage Docker clusters (Docker Inc., v1.0.1)

Server:
 Containers: 0
  Running: 0
  Paused: 0
  Stopped: 0
 Images: 0
 Server Version: 19.03.2-tp2
 Storage Driver: windowsfilter (windows) lcow (linux)
  Windows: 
  LCOW: 
 Logging Driver: json-file
 Plugins:
  Volume: local
  Network: ics l2bridge l2tunnel nat null overlay transparent
  Log: awslogs etwlogs fluentd gcplogs gelf json-file local logentries splunk syslog
 Swarm: inactive
 Default Isolation: process
 Kernel Version: 10.0 18362 (18362.1.amd64fre.19h1_release.190318-1202)
 Operating System: Windows Server Datacenter Version 1903 (OS Build 18362.295)
 OSType: windows
 Architecture: x86_64
 CPUs: 4
 Total Memory: 32GiB
 Name: azwin-1-519d20
 ID: 7JIE:3EHH:REBU:5GIT:Q2PF:UZSX:743O:DUMD:6HTG:G3W6:L3CK:2YM2
 Docker Root Dir: D:\docker
 Debug Mode: false
 Registry: https://index.docker.io/v1/
 Labels:
 Experimental: true
 Insecure Registries:
  10.0.0.4:5000
  127.0.0.0/8
 Registry Mirrors:
  http://10.0.0.4:5000/
 Live Restore Enabled: false
```

The "bad" news; the 1903 machines have Windows Defender enabled, which is currently preventing CI to run https://ci.docker.com/public/blue/rest/organizations/jenkins/pipelines/moby/branches/PR-39846/runs/1/nodes/240/log/?start=0

(until https://github.com/moby/moby/pull/39804 is merged, or the configuration of those machines is changed)

```

ERROR: Failed 'ERROR: Windows Defender real time protection must be disabled for integration tests' at 09/02/2019 10:44:13
 At D:\gopath\src\github.com\docker\docker\hack\ci\windows.ps1:269 char:22
 + ... defender) { Throw "ERROR: Windows Defender real time protection must  ...
 +                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

thaJeztah commented:
@StefanScherer ^^ looks like we may want to update the configuration for those machines (or explicitly disable it through our Jenkinsfile?)

kolyshkin reviewed (approved):
SGTM

thaJeztah commented on hack/ci/windows.ps1:
```diff
@@ -960,6 +960,9 @@ Catch [Exception] {
 }
 Finally {
+    # Preserve the LastExitCode of the tests
+    $tmpLastExitCode = $LastExitCode
```
@StefanScherer this was just an attempt to see if this would help with the RS1 builds sometimes not exiting; not sure if this would help or not (happy to drop this commit if it doesn't make sense)

thaJeztah replied:
Moved this to a separate PR https://github.com/moby/moby/pull/39854

thaJeztah commented:
Failure on RS5 is a flaky test: https://github.com/moby/moby/issues/36801
https://ci.docker.com/public/blue/rest/organizations/jenkins/pipelines/moby/branches/PR-39846/runs/5/nodes/41/log/?start=0

```
[2019-09-03T08:58:49.666Z] --- FAIL: TestJSONFileLoggerWithOpts (0.07s)
[2019-09-03T08:58:49.666Z]     jsonfilelog_test.go:187: open C:\Users\ContainerAdministrator\AppData\Local\Temp\docker-logger-344571653\container.log.1: The process cannot access the file because it is being used by another process.
```

thaJeztah commented:
We might want to consider running this one in Hyper-V mode (which I think is where the linked issue occurs)

thaJeztah commented:
Pushed a commit to run Windows 1903 with hyper-v isolation

thaJeztah commented:
couple of failures now that I enabled hyper-v; https://ci.docker.com/public/job/moby/job/PR-39846/7/execution/node/197/log/?consoleFull


```
22:46:41  FAIL: docker_api_containers_test.go:2006: DockerSuite.TestContainersAPICreateMountsCreate
22:46:41  
22:46:41  case 0 - config: {volume  c:\foo false  <nil> <nil> <nil>}
22:46:41  case 1 - config: {volume  c:\foo\ false  <nil> <nil> <nil>}
22:46:41  case 2 - config: {volume test1 c:\foo false  <nil> <nil> <nil>}
22:46:41  docker_api_containers_test.go:2180:
22:46:41      poll.WaitOn(c, containerExit(apiclient, container.ID), poll.WithDelay(time.Second))
22:46:41  d:/gopath/src/github.com/docker/docker/vendor/gotest.tools/poll/poll.go:128:
22:46:41      t.Fatalf("timeout hit after %s: %s", config.Timeout, lastMessage)
22:46:41  ... Error: timeout hit after 10s: container f84c19efa123e04f676f8c3f1424c2eb700a8680f38d5529c925b3c895bd248c is running, waiting for exit
```

Looks like this test should be either skipped on hyper-v, or modified;


```
22:49:09  FAIL: docker_cli_create_test.go:302: DockerSuite.TestCreateWithWorkdir
22:49:09  
22:49:09  assertion failed: 
22:49:09  Command:  d:\CI-7\CI-f3768a669\binary\docker.exe cp foo:c:\home\foo\bar c:\tmp
22:49:09  ExitCode: 1
22:49:09  Error:    exit status 1
22:49:09  Stdout:   
22:49:09  Stderr:   Error response from daemon: filesystem operations against a running Hyper-V container are not supported
22:49:09  
22:49:09  
22:49:09  Failures:
22:49:09  ExitCode was 1 expected 0
22:49:09  Expected no error
```

thaJeztah commented:
With `--storage-opt size=xxx`, building the BusyBox image now fails, which may be the issue reported in https://github.com/docker/for-win/issues/3884#issuecomment-510973455


https://ci.docker.com/public/job/moby/job/PR-39846/11/execution/node/240/log/


```
17:35:13  Sending build context to Docker daemon   2.56kB
17:35:13  
17:35:13  Step 1/6 : FROM microsoft/windowsservercore
17:35:13   ---> 55b4d18a7eea
17:35:13  Step 2/6 : RUN mkdir C:\tmp && mkdir C:\bin
17:35:13   ---> Running in 86b2979eda7a
17:35:31  powershell.exe : container 86b2979eda7a683b90bd96599ae8e1d64c71a06aab1690c2634cd863fbad2c1c encountered an error during Start: failure in a Windows system call: The virtual machine or 
17:35:31  container exited unexpectedly. (0xc0370106)
17:35:31  At D:\gopath\src\github.com\docker\docker@tmp\durable-9e3ff7cb\powershellWrapper.ps1:3 char:1
17:35:31  + & powershell -NoProfile -NonInteractive -ExecutionPolicy Bypass -Comm ...
17:35:31  + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
17:35:31      + CategoryInfo          : NotSpecified: (container 86b29...y. (0xc0370106):String) [], RemoteException
17:35:31      + FullyQualifiedErrorId : NativeCommandError
17:35:31   
17:35:31  
17:35:31  
17:35:31  ERROR: Failed 'ERROR: Failed to build busybox image' at 09/04/2019 15:35:28
17:35:31  At D:\gopath\src\github.com\docker\docker\hack\ci\windows.ps1:786 char:21
17:35:31  +                     Throw "ERROR: Failed to build busybox image"
17:35:31  +                     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
17:35:31 
```

thaJeztah commented:
Added a commit to _only_ set the size for build-containers to 127GB if no size was configured on the daemon.

thaJeztah commented:
Aaaaand... unit tests failed on https://github.com/moby/moby/pull/39856

```

[2019-09-05T15:03:42.298Z] --- FAIL: TestJSONFileLoggerWithOpts (0.01s)
[2019-09-05T15:03:42.298Z]     jsonfilelog_test.go:187: open C:\Users\ContainerAdministrator\AppData\Local\Temp\docker-logger-171753449\container.log.1: The process cannot access the file because it is being used by another process.
[2019-09-05T15:03:42.298Z] FAIL
[2019-09-05T15:03:42.298Z] coverage: 62.1% of statements
[2019-09-05T15:03:42.298Z] FAIL	github.com/docker/docker/daemon/logger/jsonfilelog	0.410s
```

ddebroy commented:
cc @vikramhh and @ameyag

thaJeztah commented:
Ok, that last commit didn't help it seems;

```
[2019-09-05T15:32:58.969Z] Sending build context to Docker daemon   2.56kB
[2019-09-05T15:32:58.969Z] 
[2019-09-05T15:32:58.969Z] Step 1/6 : FROM microsoft/windowsservercore
[2019-09-05T15:32:58.969Z]  ---> 55b4d18a7eea
[2019-09-05T15:32:58.969Z] Step 2/6 : RUN mkdir C:\tmp && mkdir C:\bin
[2019-09-05T15:32:59.430Z]  ---> Running in cfddf17ac863
[2019-09-05T15:33:14.283Z] powershell.exe : container cfddf17ac863206c89e1d3c19b427401884533c7642a2d50d1db5cea2a1b8647 encountered an error during Start: failure in a Windows system call: The virtual machine or 
[2019-09-05T15:33:14.283Z] container exited unexpectedly. (0xc0370106)
[2019-09-05T15:33:14.283Z] At D:\gopath\src\github.com\docker\docker@tmp\durable-dec09710\powershellWrapper.ps1:3 char:1
[2019-09-05T15:33:14.283Z] + & powershell -NoProfile -NonInteractive -ExecutionPolicy Bypass -Comm ...
[2019-09-05T15:33:14.283Z] + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[2019-09-05T15:33:14.283Z]     + CategoryInfo          : NotSpecified: (container cfddf...y. (0xc0370106):String) [], RemoteException
[2019-09-05T15:33:14.283Z]     + FullyQualifiedErrorId : NativeCommandError
[2019-09-05T15:33:14.283Z]  
```

Can try disabling that line altogether (perhaps at that point the config isn't set)

thaJeztah commented:
Still failing; removing the `DOCKER_STORAGE_OPT=size=40G` again to see if it passes then

thaJeztah commented:
Ok, so removing the `size` option in both places makes it work; https://ci.docker.com/public/job/moby/job/PR-39846/17/execution/node/166/log/


```
19:20:10  Sending build context to Docker daemon   2.56kB
19:20:10  
19:20:11  Step 1/6 : FROM microsoft/windowsservercore
19:20:11   ---> 55b4d18a7eea
19:20:11  Step 2/6 : RUN mkdir C:\tmp && mkdir C:\bin
19:20:11   ---> Running in ab87ba4cd49b
19:20:36  Removing intermediate container ab87ba4cd49b
19:20:36   ---> d90939ced1b8
19:20:36  Step 3/6 : ADD http://frippery.org/files/busybox/busybox.exe /bin/
19:20:36  
19:20:36   ---> 6dd758a8e2dd
19:20:36  Step 4/6 : RUN setx /M PATH "C:\bin;%PATH%"
19:20:36   ---> Running in 89b15b505d17
19:20:40  
19:20:40  SUCCESS: Specified value was saved.
19:20:43  Removing intermediate container 89b15b505d17
19:20:43   ---> 6c66ae6bcbb2
19:20:43  Step 5/6 : RUN powershell busybox.exe --list ^|%{$nul = cmd /c mklink C:\bin\$_.exe busybox.exe}
19:20:43   ---> Running in fca54fa3e1a0
19:21:00  [91mCannot create a file when that file already exists.
19:21:11  [0mRemoving intermediate container fca54fa3e1a0
19:21:11   ---> 0b7e7b4e58b4
19:21:11  Step 6/6 : CMD ["sh"]
19:21:11   ---> Running in bfaf81f62a13
19:21:11  Removing intermediate container bfaf81f62a13
19:21:11   ---> 2b7dd8945680
19:21:11  Successfully built 2b7dd8945680
19:21:11  Successfully tagged busybox:latest
19:21:11  INFO: Docker images of the daemon under test
```

thaJeztah commented:
Rebased this on top of https://github.com/moby/moby/pull/40128, and triggered a build with all Windows variants enabled

thaJeztah commented:
Seeing a lot of these failures on RS1;

```

[2019-10-30T14:22:37.513Z]     --- FAIL: TestDockerSuite/TestEventsContainerEventsAttrSort (3.17s)
[2019-10-30T14:22:37.513Z]         cli.go:29: assertion failed: 
[2019-10-30T14:22:37.513Z]             Command:  d:\CI\PR-39846\32\binary\docker.exe run --rm --name container-events-test busybox true
[2019-10-30T14:22:37.513Z]             ExitCode: 125
[2019-10-30T14:22:37.513Z]             Error:    exit status 125
[2019-10-30T14:22:37.513Z]             Stdout:   
[2019-10-30T14:22:37.513Z]             Stderr:   d:\CI\PR-39846\32\binary\docker.exe: Error response from daemon: process 1176 in container 01ef1c1ff2fddc4f76fdeb147bf724ce80a00bf46f90b4e3edf51531cf818426 encountered an error during hcsshim::Process::StdioLegacy: failure in a Windows system call: Element not found. (0x490).
[2019-10-30T14:22:37.513Z]             
[2019-10-30T14:22:37.513Z]             
[2019-10-30T14:22:37.513Z]             Failures:
[2019-10-30T14:22:37.513Z]             ExitCode was 125 expected 0
[2019-10-30T14:22:37.513Z]             Expected no error

```

thaJeztah commented:
Ha! So windows 1903 / Hyper-V is green on this one, and RS1 and RS5 are red 🤷‍♂

thaJeztah commented:
Removed commits that were not strictly needed; this still runs Windows 1903 with Hyper-V enabled by default (if needed we can add an additional stage with 1903 running with process isolation)


@kolyshkin @StefanScherer PTAL

StefanScherer reviewed (approved):
LGTM

`TESTRUN_DIR` is set to `CI-$BUILD_NUMBER` compared to RS5, but this should be fine.

thaJeztah commented:
ah, yes this PR was created quite a while back, so think I didn't sync that; I can update to get it right

StefanScherer commented:
Thanks, the fixed "CI" folder was to easier remove all old folders in drive:\CI from previous builds IIRC.

thaJeztah committed:
Jenkinsfile: windows-1903: send junit.xml results

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Looks like autogen is now failing; failure looks similar to the problem we had on RS5, and may be because the Windows base images are updated with the latest security fixes, and apparently broke things;

```

[2020-02-17T18:04:04.062Z] INFO: make.ps1 ended at 02/17/2020 18:04:04
[2020-02-17T18:04:05.034Z] powershell.exe : .go-autogen.ps1 failed with error Failed to compile event message resources
[2020-02-17T18:04:05.034Z] At D:\gopath\src\github.com\docker\docker@tmp\durable-98f5b357\powershellWrapper.ps1:3 char:1
[2020-02-17T18:04:05.034Z] + & powershell -NoProfile -NonInteractive -ExecutionPolicy Bypass -Comm ...
[2020-02-17T18:04:05.034Z] + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[2020-02-17T18:04:05.034Z]     + CategoryInfo          : NotSpecified: (.go-autogen.ps1...ssage resources:String) [], RemoteException
[2020-02-17T18:04:05.034Z]     + FullyQualifiedErrorId : NativeCommandError
[2020-02-17T18:04:05.034Z]  
[2020-02-17T18:04:05.034Z] At C:\gopath\src\github.com\docker\docker\hack\make\.go-autogen.ps1:83 char:5
[2020-02-17T18:04:05.034Z] +     Throw ".go-autogen.ps1 failed with error $_"
[2020-02-17T18:04:05.034Z] 
[2020-02-17T18:04:05.034Z] +     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[2020-02-17T18:04:05.034Z] 
[2020-02-17T18:04:05.034Z]     + CategoryInfo          : OperationStopped: (.go-autogen.ps1...ssage resou 
[2020-02-17T18:04:05.034Z]    rces:String) [], RuntimeException
[2020-02-17T18:04:05.034Z] 
[2020-02-17T18:04:05.034Z]     + FullyQualifiedErrorId : .go-autogen.ps1 failed with error Failed to comp 
[2020-02-17T18:04:05.034Z]    ile event message resources
[2020-02-17T18:04:05.034Z]  
[2020-02-17T18:04:05.034Z] 
[2020-02-17T18:04:05.496Z] 
[2020-02-17T18:04:05.496Z] 
[2020-02-17T18:04:05.496Z] ERROR: Failed 'ERROR: Failed to build binary' at 02/17/2020 18:04:05
[2020-02-17T18:04:05.496Z] At D:\gopath\src\github.com\docker\docker\hack\ci\windows.ps1:516 char:13
[2020-02-17T18:04:05.496Z] +             Throw "ERROR: Failed to build binary"
[2020-02-17T18:04:05.496Z] +             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[2020-02-17T18:04:05.496Z] 
```

StefanScherer commented:
I've rolled back the 1903 agent to the previous patch level, and restarted Jenkins. Let's see how that works (but we still use the latest container image, so it might still have problems).

StefanScherer commented:
@thaJeztah I've triggered another build with the latest 1903 VM image. The 1903 build is 💚, 
RS5 is failing with the workaround applied (using the older base image), but this should be fixed with #40512.

thaJeztah commented:
Ah, nice; thanks!

thaJeztah commented:
Rebased; if all is well, this should be green now 🤞

StefanScherer commented:
at least green for RS5 and 1903 :-)

thaJeztah commented:
rebased again to get a fresh run of CI

@tiborvass PTAL

thaJeztah commented:
Rebased (again)

olljanat commented:
I think that it is good idea to go directly with 2004 and if that passes build then this probably should be merged?

thaJeztah commented:
> I think that it is good idea to go directly with 2004 and if that passes build then this probably should be merged?

@StefanScherer do we have 2004 nodes in CI?

StefanScherer commented:
@thaJeztah I thought the same this morning. :-) Not yet, but that should be easy to add. I'll look at it this week and report back.

StefanScherer commented:
@thaJeztah Just added Windows Server, version 2004 instances with label `windows-2004`. So it's good to update to latest semi-annual version.

thaJeztah committed:
Jenkinsfile: update to Windows 2022

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Thanks! Updated to Windows 2004

StefanScherer commented:
Two failing tests on Windows Server 2004

```
[2020-09-15T19:57:24.225Z] === Failed
[2020-09-15T19:57:24.225Z] === FAIL: github.com/docker/docker/pkg/archive TestChangesDirsEmpty (0.12s)
[2020-09-15T19:57:24.225Z]     changes_test.go:261: Reported changes for identical dirs: [{\dirSymlink C}]
[2020-09-15T19:57:24.225Z] 
[2020-09-15T19:57:24.225Z] === FAIL: github.com/docker/docker/pkg/archive TestChangesDirsMutated (0.10s)
[2020-09-15T19:57:24.225Z]     changes_test.go:391: unexpected change "C \\dirSymlink" "\\dirnew"
[2020-09-15T19:57:24.225Z] 
```

But the Jenkins pipeline reports "green". 🤔 
Still a success to have feedback from latest semi-annual release.

thaJeztah commented:
> Two failing tests on Windows Server 2004

Interesting

thaJeztah commented:
So the `copydir` function is simply shelling out to `robocopy ` - would there be a change in that in Windows 2004?
https://github.com/moby/moby/blob/d96f61cfe1430791144b5b3b4f07aea0fa19e8eb/pkg/archive/changes_test.go#L32-L37

thaJeztah commented:
Some report here of it being borked in Windows 2004; https://answers.microsoft.com/en-us/windows/forum/all/robocopy-broken-in-win-10-v2004/71001e3e-f4e3-4c0f-886a-2ac16cc0ccd5 and https://windowsserver.uservoice.com/forums/295047-general-feedback/suggestions/40946653-robocopy-windows-10-2004-19041-388-not-removes

olljanat commented:
> Two failing tests on Windows Server 2004
> ...
> But the Jenkins pipeline reports "green". 🤔

Log says:
```INFO: Unit tests results(C:\windows\TEMP\junit-report-unit-tests.xml) exist. pwd=D:\gopath\src\github.com\docker\docker```

Is there someway to actually see what is on that XML report? It status reported wrong already on it or is that just parsed incorrectly after tests?

thaJeztah commented:
I think the XML should be uploaded as artifact (not at my computer right now); https://ci-next.docker.com/public/blue/organizations/jenkins/moby/detail/PR-39846/55/artifacts

olljanat commented on Jenkinsfile:
```diff
@@ -1091,6 +1099,67 @@ pipeline {
+                    }
+                    post {
+                        always {
+                            junit testResults: 'bundles/junit-report-*.xml', allowEmptyResults: true
```
This one is looking for XML report from bundles but based on log it is created to C:\windows\TEMP\junit-report-unit-tests.xml Also we probably should not allow empty result.

olljanat commented on Jenkinsfile:
```diff
@@ -1091,6 +1099,67 @@ pipeline {
+                                Write-Host -ForegroundColor Green "Creating ${bundleName}-bundles.zip"
+
+                                # archiveArtifacts does not support env-vars to , so save the artifacts in a fixed location
+                                Compress-Archive -Path "${env:TEMP}/CIDUT.out", "${env:TEMP}/CIDUT.err", "${env:TEMP}/junit-report-*.xml" -CompressionLevel Optimal -DestinationPath "${bundleName}-bundles.zip"
```
`${env:TEMP}/junit-report-*.xml` looks to be correct way to look those files as junit-report-unit-tests.xml ended up to artifacts and it contains right result:
```xml
<testcase classname="github.com/docker/docker/pkg/archive" name="TestChangesDirsEmpty" time="0.120000">
	<failure message="Failed" type="">=== RUN   TestChangesDirsEmpty&#xA;--- FAIL: TestChangesDirsEmpty (0.12s)&#xA;    changes_test.go:261: Reported changes for identical dirs: [{\dirSymlink C}]&#xA;</failure>
</testcase>
<testcase classname="github.com/docker/docker/pkg/archive" name="TestChangesDirsMutated" time="0.100000">
	<failure message="Failed" type="">=== RUN   TestChangesDirsMutated&#xA;--- FAIL: TestChangesDirsMutated (0.10s)&#xA;    changes_test.go:391: unexpected change &#34;C \\dirSymlink&#34; &#34;\\dirnew&#34;&#xA;</failure>
</testcase>
```

olljanat reviewed (commented):
Looks to be broken already on RS1 and RS5 too. Added inline comments and testing possible solution on #40599

olljanat commented:
@thaJeztah looks that I figured it out. Can you try include https://github.com/moby/moby/pull/40599/commits/c799853b23d37cdcae61e09e1439fe8a69f2624e to here and do same changes for 2004 Jenkinsfile as it looks that those changes does not take effect on my PR (probably need merge to master first).

olljanat commented:
@thaJeztah maybe it would be good idea to include https://github.com/moby/moby/blob/82b5ff8026506017a3b01f7b6e0ece5c226a56bb/Jenkinsfile#L1079 for 2004 too and see that those broken unit tests actually make CI fail?

thaJeztah commented:
Rebased, and included the changes from https://github.com/moby/moby/pull/40599 in the new stage in the Jenkinsfile

olljanat commented:
@thaJeztah afaiu we need change also https://github.com/moby/moby/blob/89bec2e26753a6ab9248835b7d663d977a4e324b/hack/ci/windows.ps1#L1012 to `exit 1` because on PowerShell throw command actually exist with error code 0 and that why Jenkins does not catch those failing unit tests.

StefanScherer commented:
@olljanat I think the `if` block https://github.com/moby/moby/blob/89bec2e26753a6ab9248835b7d663d977a4e324b/hack/ci/windows.ps1#L1007-L1011 before that `throw` tries exactly that. 🤔

olljanat commented:
@StefanScherer true. So many hacks on this one that I always lost track.... Anyway, based on log it will also run `Finally {}` section so most probably it goes wrong on this one https://github.com/moby/moby/blob/89bec2e26753a6ab9248835b7d663d977a4e324b/hack/ci/windows.ps1#L1016

olljanat commented:
> Some report here of it being borked in Windows 2004; https://answers.microsoft.com/en-us/windows/forum/all/robocopy-broken-in-win-10-v2004/71001e3e-f4e3-4c0f-886a-2ac16cc0ccd5 and https://windowsserver.uservoice.com/forums/295047-general-feedback/suggestions/40946653-robocopy-windows-10-2004-19041-388-not-removes

@thaJeztah @StefanScherer can we get Windows Server 2022 preview to CI to see if this issue have been already fixed on that one? https://cloudblogs.microsoft.com/windowsserver/2021/03/02/announcing-windows-server-2022-now-in-preview/

StefanScherer commented:
@olljanat Yes I think that's possible. There is an Azure VM base image. I can take a look.

StefanScherer commented:
@olljanat @thaJeztah we now have `windows-2022` Nodes in Jenkins! Maybe it‘s time to update the PR once again 😅

thaJeztah commented:
Updated to 2022 🤞

StefanScherer commented:
Oh yes, MSFT hasn't created a "2022" tag 😢
`Error response from daemon: manifest for mcr.microsoft.com/windows/servercore:2022 not found: manifest unknown: manifest tagged by "2022" is not found`

The `docker info` shows
```
  Kernel Version: 10.0 20295 (20295.1.amd64fre.fe_release.210212-1600)
  Operating System: Windows Server 2022 Datacenter Version 2009 (OS Build 20295.1)
```


Ah, but there are matching insider images with tag `10.0.20295.1`.

- `mcr.microsoft.com/windows/nanoserver/insider:10.0.20295.1`
- `mcr.microsoft.com/windows/servercore/insider:10.0.20295.1`

thaJeztah committed:
Jenkinsfile: temporarily use insider build for Windows 2022 (SAC)

Images for Windows 2022 (SAC) are not yet available, so using insider builds
in the meantime; mcr.microsoft.com/windows/servercore/insider:10.0.20295.1

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Pushed a commit to (temporarily) use `mcr.microsoft.com/windows/servercore/insider:10.0.20295.1`

StefanScherer commented:
Thanks. Ok we have two failing unit tests 

```

[2021-03-24T00:18:28.786Z] === FAIL: github.com/docker/docker/pkg/archive TestChangesDirsEmpty (0.21s)
[2021-03-24T00:18:28.786Z]     changes_test.go:261: Reported changes for identical dirs: [{\dirSymlink C}]
[2021-03-24T00:18:28.786Z] 
[2021-03-24T00:18:28.786Z] === FAIL: github.com/docker/docker/pkg/archive TestChangesDirsMutated (0.14s)
[2021-03-24T00:18:28.786Z]     changes_test.go:391: unexpected change "C \\dirSymlink" "\\dirnew"
```

thaJeztah commented:
Ah! Yes, look like the same tests we found originally in windows 2004? https://github.com/moby/moby/pull/39846#issuecomment-692949972

StefanScherer commented:
Oh yes, indeed. I didn't look up to the older comments 😅 
We should log stdout/stderr in case of an error.
In the meantime I manually tried the `robocopy` command with the sample dir created from the test code in this insider container.

```
C:\>robocopy src dst3 /SL /COPYALL /MIR

-------------------------------------------------------------------------------
   ROBOCOPY     ::     Robust File Copy for Windows
-------------------------------------------------------------------------------

  Started : Wednesday, March 24, 2021 10:09:12 AM
   Source : C:\src\
     Dest : C:\dst3\

    Files : *.*

  Options : *.* /S /E /COPYALL /PURGE /MIR /R:1000000 /W:30

------------------------------------------------------------------------------

          New Dir         12    C:\src\
            New File                   0        dirSymlink
100%        New File                   6        file1
100%        New File                   6        file2
100%        New File                   6        file3
100%        New File                   6        file4
100%        New File                   6        file5
100%        New File                   6        file6
100%        New File                   6        file7
            New File                   0        symlink1
            New File                   0        symlink2
            New File                   0        symlink3
            New File                   0        symlink4
          New Dir          2    C:\src\dir1\
100%        New File                   8        file1-1
100%        New File                   8        file1-2
          New Dir          2    C:\src\dir2\
100%        New File                   8        file2-1
100%        New File                   8        file2-2
          New Dir          2    C:\src\dir3\
100%        New File                   8        file3-1
100%        New File                   8        file3-2
          New Dir          2    C:\src\dir4\
100%        New File                   8        file3-1
100%        New File                   8        file3-2

------------------------------------------------------------------------------

               Total    Copied   Skipped  Mismatch    FAILED    Extras
    Dirs :         5         5         0         0         0         0
   Files :        20        20         0         0         0         0
   Bytes :       106       106         0         0         0         0
   Times :   0:00:00   0:00:00                       0:00:00   0:00:00


   Speed :               6,235 Bytes/sec.
   Speed :               0.357 MegaBytes/min.
   Ended : Wednesday, March 24, 2021 10:09:12 AM


C:\>echo %ERRORLEVEL%
1
```

Maybe we should just skip the error code? 🤔

olljanat commented:
Error code 1 is normal behavior for Robocopy https://docs.microsoft.com/en-us/troubleshoot/windows-server/backup-and-storage/return-codes-used-robocopy-utility

I think that we hit this one https://windowsserver.uservoice.com/forums/295047-general-feedback/suggestions/40946653-robocopy-windows-10-2004-19041-388-not-removes which would mean that something is changed either on robocopy or NTFS. Just guessing that maybe it can be related to those changes which was done because of WSL? https://github.com/golang/go/issues/32088#issuecomment-502850674

@cpuguy83 can you ask someone from Windows team to check this one?

thaJeztah commented:
I think I'll add a commit to conditionally skip these tests, so that we can unblock this PR (it's useful to have this version of Windows running in CI)

After that, I'll open a draft PR that reverts that commit, so that we can use that for discussing/testing fixes

thaJeztah committed:
pkg/archive temporarily skip some tests on Windowsd V19H1 (1903) and up

These tests fail, possibly due to changes in the kernel. Temporarily skipping
these tests, so that we at least have some coverage on these windows versions
in this repo, and we can look into this specific issue separately.;

    === FAIL: github.com/docker/docker/pkg/archive TestChangesDirsEmpty (0.21s)
        changes_test.go:261: Reported changes for identical dirs: [{\dirSymlink C}]

    === FAIL: github.com/docker/docker/pkg/archive TestChangesDirsMutated (0.14s)
        changes_test.go:391: unexpected change "C \\dirSymlink" "\\dirnew"

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Pushed a commit to skip those tests 🤞

olljanat commented:
CI green from relevant parts so LGTM

> After that, I'll open a draft PR that reverts that commit, so that we can use that for discussing/testing fixes

When you do so can you trigger it without Hyper-V (process isolation) to be able see if we have new issues on that combination.

olljanat commented on Jenkinsfile:
```diff
@@ -1115,6 +1116,8 @@ pipeline {
                         DOCKER_BUILDKIT        = '0'
                         DOCKER_DUT_DEBUG       = '1'
+                        DOCKER_DUT_HYPERV      = '1'
+                        DOCKER_STORAGE_OPTS    = 'size=40G'
```
I guess that these settings should be under win-2022 section?

thaJeztah replied:
Oh, wow, yes, you are right. Either has been a bad rebase at some point, or I may have been testing at the time if the "size" issue also reproduced on RS5 in Hyper-V

thaJeztah commented:
> EDIT: I just noticed that 2022 is actually running on process isolation mode now and RS5 on Hyper-V isolation (which why it fails) so probably it would be better to skip that part from this PR too.

Fixed; moved it to the win-2022 part, and (to test) reverted the last (skip tests) commit; I think it's good to have _at least one_ of the Windows stages run with Hyper-V; that said, we discussed removing RS1 from the list at some point; perhaps we can add a "windows latest (Hyper-V)" and "windows latest (Process)".
It's a pity that Jenkins declarative pipeline doesn't (I think still?) allow further nesting, otherwise we could "share" the `build` step, which takes quite a while on Windows, and after that branch to process/hyper-v for tests.

thaJeztah commented:
Hmmm.... looks like running it on Hyper-V fails; getting some hcsshim failure


```
hcsshim::CreateComputeSystem a9d9068c72032118b74834d50c4205ac6c23bd506caee08b0d2d01356413cc28: The request is not supported.
```

```
[2021-03-25T16:26:00.394Z] INFO: Building busybox
[2021-03-25T16:26:00.394Z] Sending build context to Docker daemon   5.12kB
[2021-03-25T16:26:00.394Z] 
[2021-03-25T16:26:00.394Z] Step 1/13 : ARG WINDOWS_BASE_IMAGE=mcr.microsoft.com/windows/servercore
[2021-03-25T16:26:00.394Z] Step 2/13 : ARG WINDOWS_BASE_IMAGE_TAG=ltsc2019
[2021-03-25T16:26:00.394Z] Step 3/13 : ARG BUSYBOX_VERSION=FRP-3329-gcf0fa4d13
[2021-03-25T16:26:00.394Z] Step 4/13 : ARG BUSYBOX_SHA256SUM=bfaeb88638e580fc522a68e69072e305308f9747563e51fa085eec60ca39a5ae
[2021-03-25T16:26:00.394Z] Step 5/13 : FROM ${WINDOWS_BASE_IMAGE}:${WINDOWS_BASE_IMAGE_TAG}
[2021-03-25T16:26:00.394Z]  ---> 39d157a84080
[2021-03-25T16:26:00.394Z] Step 6/13 : RUN mkdir C:\tmp && mkdir C:\bin
[2021-03-25T16:26:00.394Z]  ---> Running in a9d9068c7203
[2021-03-25T16:26:00.867Z] powershell.exe : hcsshim::CreateComputeSystem a9d9068c72032118b74834d50c4205ac6c23bd506caee08b0d2d01356413cc28: The request is not supported.
[2021-03-25T16:26:00.867Z] At D:\gopath\src\github.com\docker\docker@tmp\durable-ac69d9f3\powershellWrapper.ps1:3 char:1
[2021-03-25T16:26:00.867Z] + & powershell -NoProfile -NonInteractive -ExecutionPolicy Bypass -Comm ...
[2021-03-25T16:26:00.867Z] + ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
[2021-03-25T16:26:00.867Z]     + CategoryInfo          : NotSpecified: (hcsshim::Create... not supported.:String) [], RemoteException
[2021-03-25T16:26:00.867Z]     + FullyQualifiedErrorId : NativeCommandError
[2021-03-25T16:26:00.867Z]  
[2021-03-25T16:26:00.867Z] (extra info: {"SystemType":"Container","Name":"a9d9068c72032118b74834d50c4205ac6c23bd506caee08b0d2d01356413cc28","Owner":"docker","IgnoreFlushesDuringBoot":true,"LayerFolderPath":"D:\\CI\\PR-3
[2021-03-25T16:26:00.867Z] 9846\\63\\daemon\\windowsfilter\\a9d9068c72032118b74834d50c4205ac6c23bd506caee08b0d2d01356413cc28","Layers":[{"ID":"11864a36-109a-522a-be2d-24a868e1cb10","Path":"D:\\CI\\PR-39846\\63\\daemon\\
[2021-03-25T16:26:00.867Z] windowsfilter\\1f640e261b425edcdf3c71f650e2540dbc080cd9c23efc5fed2f42e69553198b"}],"HostName":"a9d9068c7203","HvPartition":true,"EndpointList":["38084759-8286-449b-a6ef-f93f346a6407"],"HvRunti
[2021-03-25T16:26:00.867Z] me":{"ImagePath":"D:\\CI\\PR-39846\\63\\daemon\\windowsfilter\\1f640e261b425edcdf3c71f650e2540dbc080cd9c23efc5fed2f42e69553198b\\UtilityVM"},"AllowUnqualifiedDNSQuery":true})
[2021-03-25T16:26:00.867Z] 
[2021-03-25T16:26:00.867Z] 
[2021-03-25T16:26:00.867Z] ERROR: Failed 'ERROR: Failed to build busybox image' at 03/25/2021 16:26:00
[2021-03-25T16:26:00.867Z] At D:\gopath\src\github.com\docker\docker\hack\ci\windows.ps1:824 char:17
[2021-03-25T16:26:00.867Z] +                 Throw "ERROR: Failed to build busybox image"
[2021-03-25T16:26:00.867Z] +                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

olljanat commented:
@thaJeztah as those tests already passed without Hyper-V maybe it would be better to disable it again, squash commits and merge this one? I was hoping to be able use 2022 on #41479 (testing my proposal https://github.com/moby/moby/issues/41455#issuecomment-806340011) but it looks to be still that Jenkinsfile changes does not take effect before merging to master.

> I think it's good to have _at least one_ of the Windows stages run with Hyper-V; that said, we discussed removing RS1 from the list at some point; perhaps we can add a "windows latest (Hyper-V)" and "windows latest (Process)".

Honestly I do not care about Hyper-V mode as we do not use it anywhere but sure because that feature exist then it probably should be part of tests.

thaJeztah commented:
Removed the commits that (I think) were all Hyper-V related; these commits were removed:

- switched it to use hyper-v
- set storage-opt: size
- disabled defender
- skipped tests

I kept those in a separate branch, so if some need to be added back I can add them; let's see what CI says

thaJeztah commented:
Pushed the other branch as draft https://github.com/moby/moby/pull/42277

olljanat commented:
@thaJeztah looks that you skipped one change too much. We need https://github.com/moby/moby/pull/42277/commits/48d0152796e1b4dc8ad487508234bffbf0c40173 here.

thaJeztah commented:
> @thaJeztah looks that you skipped one change too much. We need [48d0152](https://github.com/moby/moby/commit/48d0152796e1b4dc8ad487508234bffbf0c40173) here.

Ah! That was the one I wasn't sure about (didn't recall if it was version only, or hyper-v specific); will fix

olljanat reviewed (approved):
LGTM

StefanScherer reviewed (approved):
Still LGTM !

thaJeztah commented:
Thanks! I'll go ahead and merge 🤗
